test/libyul/Parser.cpp

```diff
@@ -642,6 +642,164 @@ BOOST_AUTO_TEST_CASE(customSourceLocations_reference_original_sloc)
 	CHECK_LOCATION(varDecl.debugData->location, "", 10, 20);
 }
 
+BOOST_AUTO_TEST_CASE(customSourceLocations_with_code_snippets)
+{
+	ErrorList errorList;
+	ErrorReporter reporter(errorList);
+	auto const sourceText = R"~~~(
+		{
+			/// @src 0:149:156  "new C(\"123\")"
+			let x := 123
+
+			let y := /** @src 1:96:165  "contract D {..." */ 128
+		}
+	)~~~";
+	EVMDialectTyped const& dialect = EVMDialectTyped::instance(EVMVersion{});
+	shared_ptr<Block> result = parse(sourceText, dialect, reporter);
+	BOOST_REQUIRE(!!result && errorList.size() == 0);
+	BOOST_REQUIRE_EQUAL(result->statements.size(), 2);
+
+	BOOST_REQUIRE(holds_alternative<VariableDeclaration>(result->statements.at(0)));
+	VariableDeclaration const& varX = get<VariableDeclaration>(result->statements.at(0));
+	CHECK_LOCATION(varX.debugData->location, "source0", 149, 156);
+
+	BOOST_REQUIRE(holds_alternative<VariableDeclaration>(result->statements.at(1)));
+	VariableDeclaration const& varY = get<VariableDeclaration>(result->statements.at(1));
+	BOOST_REQUIRE(!!varY.value);
+	BOOST_REQUIRE(holds_alternative<Literal>(*varY.value));
+	Literal const& literal128 = get<Literal>(*varY.value);
+	CHECK_LOCATION(literal128.debugData->location, "source1", 96, 165);
+}
+
+BOOST_AUTO_TEST_CASE(customSourceLocations_with_code_snippets_empty_snippet)
+{
+	ErrorList errorList;
+	ErrorReporter reporter(errorList);
+	auto const sourceText = R"(
+		/// @src 0:111:222 ""
+		{}
+	)";
+	EVMDialectTyped const& dialect = EVMDialectTyped::instance(EVMVersion{});
+	shared_ptr<Block> result = parse(sourceText, dialect, reporter);
+	BOOST_REQUIRE(!!result && errorList.size() == 0);
+	CHECK_LOCATION(result->debugData->location, "source0", 111, 222);
+}
+
+BOOST_AUTO_TEST_CASE(customSourceLocations_with_code_snippets_no_whitespace_before_snippet)
+{
+	ErrorList errorList;
+	ErrorReporter reporter(errorList);
+	auto const sourceText = R"(
+		/// @src 0:111:222"abc" def
+		{}
+	)";
+	EVMDialectTyped const& dialect = EVMDialectTyped::instance(EVMVersion{});
+	shared_ptr<Block> result = parse(sourceText, dialect, reporter);
+	BOOST_REQUIRE(!!result);
+	BOOST_REQUIRE(errorList.size() == 1);
+	BOOST_TEST(errorList[0]->type() == Error::Type::SyntaxError);
+	BOOST_TEST(errorList[0]->errorId() == 8387_error);
+	CHECK_LOCATION(result->debugData->location, "", -1, -1);
+}
+
+BOOST_AUTO_TEST_CASE(customSourceLocations_with_code_snippets_no_whitespace_after_snippet)
+{
+	ErrorList errorList;
+	ErrorReporter reporter(errorList);
+	auto const sourceText = R"(
+		/// @src 0:111:222 "abc"def
+		{}
+	)";
+	EVMDialectTyped const& dialect = EVMDialectTyped::instance(EVMVersion{});
+	shared_ptr<Block> result = parse(sourceText, dialect, reporter);
+	BOOST_REQUIRE(!!result && errorList.size() == 0);
+	CHECK_LOCATION(result->debugData->location, "source0", 111, 222);
```

Comment:
The regex for location strictly requires whitespace after the numbers. I did not change that aspect but for the quoted snippet, which is a new element, I made it more lenient - you don't have to add whitespace after it. Is that ok?

Reply:
From @chriseth:
> of course no strict requirement for whitespace after something
> whitespace is only required to separate stuff

I'm leaving it as is then.

Reply:
Actually, I think it wasn't clear from my question on the chat that I mean requiring whitespace in the case above, i.e. `"abc" def` and not requiring trailing whitespace after the quote . So I guess I should not allow this after all?

Reply:
I left it as is after all. I don't think that missing whitespace after `"` should be a syntax error and ignoring the whole tag because of it does not seem like a good idea either.